src/exercises/adapt-c/answers.md

```diff
@@ -1 +1,279 @@
+   src/exercises/adapt-c/cat/methods.c:    return (write(fildes, (const void *)(off + (uintptr_t)buf), nbyte));
+   src/exercises/adapt-c/cat/methods.c:                    if ((nw = write_off(wfd, buf, off, (size_t)nr)) < 0)
+   src/exercises/adapt-c/cat/methods.c:                            err(1, "write(2) failed");
+   ```
+   The call is in the `write_off()` function and the message with the error is
+   printed in case the call fails. Let's see what are arguments and result
+   values for the system call by setting a breakpoint for `write()` in `gdb`:
+   ```
+   # gdb ./cat-cheri
+   (...)
+   (gdb) b write
+   Function "write" not defined.
+   Make breakpoint pending on future shared library load? (y or [n]) y
+   Breakpoint 1 (write) pending.
+   (gdb) r /etc/hostid
+   Starting program: /buildroot/cat-cheri /etc/hostid
+   
+   Breakpoint 1, write (fd=1, buf=0x40810000, nbytes=37) at /source/cheribsd/lib/libc/sys/write.c:49
+   49      /source/cheribsd/lib/libc/sys/write.c: No such file or directory.
+   (gdb) info register ca0 ca1 ca2
+   ca0            0x1      0x1
+   ca1            0x40810000       0x40810000
+   ca2            0x25     0x25
+   (gdb) 
+   (gdb) disassemble 
+   Dump of assembler code for function write:
+   => 0x0000000040295dc4 <+0>:       auipc   a3,0xc2
+      0x0000000040295dc8 <+4>:       lc      ca3,1532(a3) # 0x403583c0 <_CHERI_CAPABILITY_TABLE_+2464>
+      0x0000000040295dcc <+8>:       lc      ca5,496(a3)
+      0x0000000040295dd0 <+12>:      cjr     ca5
+   End of assembler dump.
+   (gdb) 
+   ```
+   We can see that `write()` was called to write to `stdout` (`ca0`) 37 bytes
+   (`ca2`) from a buffer with an untagged capability (`ca1`). The `write()` libc
+   function does not include a trapping instruction but it jumps with `cjr`.
+   Let's see where it jumps to:
+   ```
+   (gdb) ni 4
+   _write () at _write.S:4
+   4       _write.S: No such file or directory.
+   (gdb) disassemble 
+   Dump of assembler code for function _write:
+   => 0x0000000040299130 <+0>:       li      t0,4
+      0x0000000040299132 <+2>:       ecall
+      0x0000000040299136 <+6>:       bnez    t0,0x4029913e <_write+14>
+      0x000000004029913a <+10>:      cret
+      0x000000004029913e <+14>:      auipc   t1,0xffffd
+      0x0000000040299142 <+18>:      cincoffset      ct1,ct1,-846
+      0x0000000040299146 <+22>:      cjr     ct1
+   End of assembler dump.
+   (gdb) 
+   ```
+   `write()` jumped to `_write()`, a system call wrapper written in assembly,
+   that makes a trap. Let's see what is a result of the system call:
+   ```
+   (gdb) ni 2
+   4       in _write.S
+   (gdb) info register ca0
+   ca0            0xe      0xe
+   (gdb) 
+   ```
+   The `write()` system call returned `0xe`. Looking at `errno(2)` and
+   `write(2)`, we can conclude that we passed an incorrect address to the buffer
+   which should be tagged.
```

Comment:
```suggestion
   `write(2)`, we can conclude that we passed an incorrect address to the buffer;
   here, that's likely because the capability is just the address, without a tag.
```

I wonder if it's worth introducing the term "null-derived" somewhere in this exercise?